cloud/client.go

```diff
@@ -183,27 +183,33 @@ func (c *client) doCall(ctx context.Context, method, url string, opts ...request
 
 	var status int
 	var body string
-	var err error
+	var callErr error
 	duration := time.Millisecond * 100
 	for attempt := 0; attempt < maxAttempt; attempt++ {
-		status, body, err = c.doCallImp(ctx, r, method, url, opts...)
-
-		if err != nil && strings.Contains(err.Error(), "context canceled") {
+		status, body, callErr = c.doCallImp(ctx, r, method, url, opts...)
+		if !shouldRetry(status, body, callErr, c.warnFunc) {
+			return status, body, callErr
+		}
+		if callErr != nil {
 			// Some operations can be canceled gracefully, so we can signal this to higher-level functions.
 			// Note the actual error caught here is not an instance of context.Canceled,
 			// but is an error message that contains the HTTP call plus the string "context canceled".
-			return status, body, context.Canceled
-		}
-
-		if !shouldRetry(status, body, err, c.warnFunc) {
-			return status, body, err
+			switch {
+			case strings.Contains(callErr.Error(), "context canceled"):
```

Comment:
I'm not quite following this. Are we getting `context canceled` as part of the API's response? If the HTTP client is hitting the cancellation, it should produce a legitimate `context.Canceled`, if I'm not mistaken.

Reply:
I believe that @alexcb might know the history behind looking for the string itself.

Reply:
I added the string check and left a comment above. Apparently, I wasn't getting an instance of `context.Canceled` but a an error containing that in the string. That's why I caught it here and converted to an actual `context.Canceled`

Reply:
As to why that is the case, I don't think I ever figured that out